profiling/energy-profiler/nvml/kp_power_nvml.cpp

```diff
@@ -0,0 +1,552 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+/**
+ * Kokkos NVML Power Profiler
+ * Simple Kokkos profiling tool that monitors GPU power consumption using NVML
+ * Polls nvmlDeviceGetPowerUsage() every 100ms in a background thread
+ * Calculates energy consumption by numerical integration of power over time
+ */
+
+#include <iostream>
+#include <vector>
+#include <chrono>
+#include <thread>
+#include <mutex>
+#include <atomic>
+#include <memory>
+#include <limits>
+#include <cstring>
+#include <cstdio>
+#include <cstdlib>
+#include <unistd.h>
+#include <condition_variable>
+#include <inttypes.h>
+
+// NVML header
+#include <nvml.h>
+
+// Kokkos Profiling API headers - use same approach as simple timer
+#include "kp_core.hpp"
+
+namespace KokkosTools {
+namespace NVMLPowerProfiler {
+
+// --- Global variables for NVML monitoring ---
+std::atomic<double> g_total_integrated_energy_joules(0.0);
+std::atomic<double> g_current_power_watts(0.0);
+std::atomic<bool> g_monitor_running(false);
+std::atomic<bool> g_stop_requested(false);
+
+// NVML specific globals
+std::vector<nvmlDevice_t> g_nvml_devices;
+std::vector<double> g_last_power_readings_W;
+std::unique_ptr<std::thread> g_monitoring_thread;
+std::condition_variable g_sleep_cv;
+std::mutex g_sleep_mutex;
+
+// Data collection for CSV output
+struct PowerDataPoint {
+  int64_t timestamp_ms;
+  double total_power_watts;
+  double integrated_energy_joules;
+};
+std::vector<PowerDataPoint> g_power_data_points;
```

Comment:
This is probably not the right data structure for the job.
Insertion will be costly when the capacity of the std::vector is exceeded (reallocation+copy).
Consider using std::dequeue instead.